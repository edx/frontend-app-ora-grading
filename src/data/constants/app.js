import { getConfig } from '@edx/frontend-platform';

export const routePath = `${getConfig().PUBLIC_PATH}:courseId`;
export const locationId = window.location.pathname.slice(1);
