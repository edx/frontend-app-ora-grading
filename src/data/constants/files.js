import { StrictDict } from 'utils';

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
});

export default FileTypes;
