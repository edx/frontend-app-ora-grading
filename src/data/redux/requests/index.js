export { actions, reducer } from './reducer';
