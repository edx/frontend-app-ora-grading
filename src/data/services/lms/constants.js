import { StrictDict } from 'utils';

export const lockStatuses = StrictDict({
  unlocked: 'unlocked',
  locked: 'locked',
  inProgress: 'in-progress',
});

export const gradeStatuses = StrictDict({
  ungraded: 'ungraded',
  graded: 'graded',
});

export const gradingStatuses = StrictDict({
  ungraded: gradeStatuses.ungraded,
  graded: gradeStatuses.graded,
  locked: lockStatuses.locked,
  inProgress: lockStatuses.inProgress,
});

export const gradingStatusDisplay = StrictDict({
  [gradingStatuses.ungraded]: 'Ungraded',
  [gradingStatuses.locked]: 'Currently being graded by someone else',
  [gradingStatuses.graded]: 'Grading Complete',
  [gradingStatuses.inProgress]: 'You are currently grading this response',
});

export const feedbackRequirement = StrictDict({
  disabled: 'disabled',
  required: 'required',
  optional: 'optional',
});
