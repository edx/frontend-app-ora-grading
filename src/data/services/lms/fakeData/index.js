import submissions from './submissionList';
import { mockSubmission } from './submissionFull';
import oraMetadata from './ora';
import ids from './ids';

export default {
  submissions,
  oraMetadata,
  mockSubmission,
  ids,
};
