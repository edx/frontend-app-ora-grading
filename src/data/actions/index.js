import { StrictDict } from 'utils';

import app from './app';
import grading from './grading';
import submissions from './submissions';

export default StrictDict({
  app,
  grading,
  submissions,
});
