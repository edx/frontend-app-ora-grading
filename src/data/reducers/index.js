import { combineReducers } from 'redux';

import app from './app';
import submissions from './submissions';

/* istanbul ignore next */
const rootReducer = combineReducers({
  app,
  submissions,
});

export default rootReducer;
