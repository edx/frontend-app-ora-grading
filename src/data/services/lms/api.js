import { StrictDict } from 'utils';
import { gradeStatuses, lockStatuses } from './constants';
import fakeData from './fakeData';

// import urls from './urls';
// import { pageSize, paramKeys } from './constants';
// import messages from './messages';
// import * as utils from './utils';
// const { get, post, stringifyUrl } = utils;

/*********************************************************************************
 * GET Actions
 *********************************************************************************/

const mockSuccess = (returnValFn) => (...args) => (
  new Promise((resolve) => resolve(returnValFn(...args)))
);

const mockFailure = (returnValFn) => (...args) => (
  new Promise((resolve, reject) => reject(returnValFn(...args)))
);

/**
 * get('/api/initialize', { ora_location, course_id? })
 * @return {
 *   oraMetadata: { name, prompt, type ('individual' vs 'team'), rubricConfig, fileUploadResponseConfig },
 *   courseMetadata: { courseOrg, courseName, courseNumber, courseId },
 *   submissions: {
 *     [submissionUUID]: {
 *       id: <submissionUUID>, (not currently used)
 *       username
 *       submissionUUID
 *       dateSubmitted (timestamp)
 *       gradeStatus (['ungraded', 'graded', 'locked', 'locked_by_you'?])
 *       grade: { pointsEarned, pointsPossible }
 *     },
 *     ...
 *   },
 * }
 */
const initializeApp = mockSuccess(() => ({
  oraMetadata: fakeData.oraMetadata,
  courseMetadata: fakeData.courseMetadata,
  submissions: fakeData.submissions,
}));

/**
 * get('/api/submission', { submissionUUID })
 * @return {
 *   submision: {
 *     gradeData,
 *     gradeStatus,
 *     response: { files: [{}], text: <html> },
 *   },
 * }
 */
const fetchSubmission = mockSuccess((submissionUUID) => {
  console.log({ fetchSubmission: { submissionUUID } });
  return fakeData.mockSubmission(submissionUUID);
});

/**
 * fetches the current grade, gradeStatus, and rubricResponse data for the given submission
 * get('/api/submissionStatus', { submissionUUID })
 *   @return {obj} submissionStatus object
 *   {
 *     gradeData,
 *     gradeStatus,
 *     lockStatus,
 *   }
 */
const fetchSubmissionStatus = mockSuccess((submissionUUID) => (
  fakeData.mockSubmissionStatus(submissionUUID)
));

/**
 * Fetches only the learner response for a given submission. Used for pre-fetching response
 * for neighboring submissions in the queue.
 */
export const fetchSubmissionResponse = mockSuccess((submissionUUID) => ({
  response: fakeData.mockSubmission(submissionUUID).response,
}));

/* I assume this is the "Start Grading" call, even for if a
 * submission is already graded and we are attempting re-lock.
 * Assuming the check for if allowed would happen locally first.
 * post('api/lock', { ora_location, submissionUUID });
 * @param {bool} value - new lock value
 * @param {string} submissionUUID
 */
const lockSubmission = mockSuccess(({ submissionUUID, value }) => ({
  ...fakeData.mockSubmissionStatus(submissionUUID),
  lockStatus: value ? lockStatuses.inProgress : lockStatuses.unlocked,
}));

/*
 * Assuming we do not care who has locked it or why, as there
 * is no design around communicating that info
 * post('api/lock', { submissionUUID });
 * @param {bool} value - new lock value
 * @param {string} submissionUUID
 */
const lockSubmissionFail = mockFailure(() => ({
  error: 'that did not work',
}));

/*
 * post('api/updateGrade', { submissionUUID, gradeData })
 * @param {object} gradeData - full grading submission data
 */
const updateGrade = mockSuccess((submissionUUID, gradeData) => ({
  gradeData,
  gradeStatus: gradeStatuses.graded,
  lockStatus: lockStatuses.unlocked,
}));

export default StrictDict({
  initializeApp,
  fetchSubmission,
  fetchSubmissionResponse,
  fetchSubmissionStatus,
  lockSubmission,
  lockSubmissionFail,
  updateGrade,
});
