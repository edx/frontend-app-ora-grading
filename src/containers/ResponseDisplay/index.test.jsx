import React from 'react';
import { shallow } from 'enzyme';

import createDOMPurify from 'dompurify';
import parse from 'html-react-parser';

import { fileUploadResponseOptions } from 'data/services/lms/constants';
import { selectors } from 'data/redux';

import { ResponseDisplay, mapStateToProps } from '.';

jest.mock('data/redux', () => ({
  selectors: {
    grading: {
      selected: {
        response: (state) => ({ response: state }),
      },
    },
    app: {
      ora: {
        fileUploadResponseConfig: (state) => ({ config: state }),
      },
    },
  },
}));
jest.mock('./SubmissionFiles', () => 'SubmissionFiles');
jest.mock('dompurify', () => () => ({
  sanitize: (text) => `sanitized (${text})`,
}));
jest.mock('html-react-parser', () => (text) => `parsed html (${text})`);

describe('ResponseDisplay', () => {
  describe('component', () => {
    const props = {
      response: {
        text: ['some text response here'],
        files: [
          {
            name: 'some file name.jpg',
            description: 'description for the file',
            downloadURL: '/valid-url-wink-wink',
          },
          {
            name: 'file number 2.jpg',
            description: 'description for this file',
            downloadURL: '/url-2',
          },
        ],
      },
      fileUploadResponseConfig: 'optional',
    };
    let el;
    beforeAll(() => {
      global.window = {};
    });
    beforeEach(() => {
      el = shallow(<ResponseDisplay {...props} />);
    });
    describe('snapshot', () => {
      test('file upload enable with valid response', () => {
        expect(el.instance().render()).toMatchSnapshot();
      });

      test('file upload enable without response', () => {
        el.setProps({
          response: {
            text: [],
            files: [],
          },
        });
        expect(el.instance().render()).toMatchSnapshot();
      });
      test('file upload disable with valid response', () => {
        el.setProps({
          fileUploadResponseConfig: fileUploadResponseOptions.none,
        });
        expect(el.instance().render()).toMatchSnapshot();
      });

      test('file upload disabled without response', () => {
        el.setProps({
          response: {
            text: [],
            files: [],
          },
        });
        expect(el.instance().render()).toMatchSnapshot();
      });
    });
    describe('behavior', () => {
      test('get textContents', () => {
        expect(el.instance().textContents.length).toEqual(
          props.response.text.length,
        );
        el.instance().textContents.forEach((text, index) => {
          expect(text).toEqual(
            parse(createDOMPurify(window).sanitize(props.response.text[index])),
          );
        });
      });

      test('get submittedFiles', () => {
        expect(el.instance().submittedFiles).toEqual(props.response.files);
      });
      test('get allowFileUpload', () => {
        expect(el.instance().allowFileUpload).toEqual(
          props.fileUploadResponseConfig !== fileUploadResponseOptions.none,
        );
      });
    });
  });
  describe('mapStateToProps', () => {
    let mapped;
    const testState = {
      dummyText: ['text'],
      dummyFiles: ['files', 'file-2'],
    };
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('response loads from grading.selected.response', () => {
      expect(mapped.response).toEqual(
        selectors.grading.selected.response(testState),
      );
    });
    test('response loads from grading.selected.response', () => {
      expect(mapped.fileUploadResponseConfig).toEqual(
        selectors.app.ora.fileUploadResponseConfig(testState),
      );
    });
  });
});
