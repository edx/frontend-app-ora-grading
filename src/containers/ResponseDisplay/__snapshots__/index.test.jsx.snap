// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponseDisplay component snapshot file upload disable with valid response 1`] = `
<div
  className="response-display"
>
  <Card>
    <Card.Section
      className="response-display-text-content"
    >
      parsed html (sanitized (some text response here))
    </Card.Section>
  </Card>
</div>
`;

exports[`ResponseDisplay component snapshot file upload disabled without response 1`] = `
<div
  className="response-display"
>
  <SubmissionFiles
    files={Array []}
  />
  <PreviewDisplay
    files={Array []}
  />
</div>
`;

exports[`ResponseDisplay component snapshot file upload enable with valid response 1`] = `
<div
  className="response-display"
>
  <SubmissionFiles
    files={
      Array [
        Object {
          "description": "description for the file",
          "downloadURL": "/valid-url-wink-wink",
          "name": "some file name.jpg",
        },
        Object {
          "description": "description for this file",
          "downloadURL": "/url-2",
          "name": "file number 2.jpg",
        },
      ]
    }
  />
  <PreviewDisplay
    files={
      Array [
        Object {
          "description": "description for the file",
          "downloadURL": "/valid-url-wink-wink",
          "name": "some file name.jpg",
        },
        Object {
          "description": "description for this file",
          "downloadURL": "/url-2",
          "name": "file number 2.jpg",
        },
      ]
    }
  />
  <Card>
    <Card.Section
      className="response-display-text-content"
    >
      parsed html (sanitized (some text response here))
    </Card.Section>
  </Card>
</div>
`;

exports[`ResponseDisplay component snapshot file upload enable without response 1`] = `
<div
  className="response-display"
>
  <SubmissionFiles
    files={Array []}
  />
  <PreviewDisplay
    files={Array []}
  />
</div>
`;
